### ISSUE TITLE
Allow to specify different ObjectFactory classes via annotation

### ISSUE THREAD

**Issue by ralphkar** [cucumber/cucumber-jvm#1690]
## Summary
Specifying the ObjectFactory in CucumberOptions would be highly needed.

## Description
I have all the relevant Cucumber steps injected by a Guice injector. This injector also needs all of our project specific modules (yes, we are heavily relying on Guice). So far this was ok.
But now I have to support additional features that are only available when configured. Therefore I need to create a differently configured injector for some tests. I have been looking around for a while in order to find out what is the best way to do this in Cucumber.
I found the following (old) post on StackOverflow:
https://stackoverflow.com/questions/35315963/cucumber-with-guice-multiple-guice-injector/40429150#40429150
With this approach it would be possible for me to provide different instances of InjectorSource. Unfortunately the createRuntime()method has been removed from the Cucumber runner.

## Possible Solution
Now how could I do this with the current Cucumber (I am using version 4.3.1)? Unless there is another workaround (that I am not aware of), it would be most helpful to allow to specify an ObjectFactory within the CucumberOptions annotation.

**Comment by mpkorstanje** [comment]
Currently I don't believe there is a sound solution to this problem. 

The problem is two fold. First of all the `GuiceFactory` uses properties stored in the static variable `Env.INSTANCE` to determine the name of the `InjectorSourceFactory`. It is somewhat inaccessible.

```java
    public GuiceFactory() {
        this(new InjectorSourceFactory(Env.INSTANCE).create().getInjector());
    }
```

Secondly the JavaBackend creates the object factory.

```java
    private JavaBackend(ClassFinder classFinder, TypeRegistry typeRegistry) {
        this(loadObjectFactory(classFinder, Env.INSTANCE.get(ObjectFactory.class.getName())), classFinder, typeRegistry);
    }
```

So for now you may be best forking `cucumber-junit` and hack your own needs into it. For a long term solution, I guess looking for a way to pass properties into the `GuiceFactory` rather then pulling them from a static variables would be the way to go.

**Comment by mpkorstanje** [comment]
This is going to be pretty much impossible to fix in v4. But in v5 you could implement an `ObjectFactoryProviderService` with a single method `ObjectFactory create(Properties properties);` This could then be wired up in the various runners.

Would you be willing and able to send a PR?

**Comment by ralphkar** [comment]
Yes, I could give it a try.
But, I just checked out cucumber-jvm:master but I can't even get it to work
in Eclipse.

Ralph

Am Di., 9. Juli 2019 um 09:50 Uhr schrieb M.P. Korstanje <
notifications@github.com>:

**Comment by ralphkar** [comment]
Could be because of my Java version (1.8.0_152) or Maven version (3.6.1).

Ralph

Am Di., 9. Juli 2019 um 12:43 Uhr schrieb Ralph Kar <ralph.kar@gmail.com>:

**Comment by mpkorstanje** [comment]
You'll have to do a `mvn clean install` from the CLI first to generate all the step definitions.

**Comment by ralphkar** [comment]
Yes, I've already done that. Compilation from the CLI works fine. But not in Eclipse.
Any other required steps?

**Comment by mpkorstanje** [comment]
Maybe add the generated sources folder to the build? They're in `target` somewhere.

**Comment by ralphkar** [comment]
I have started to look into develop-v5 and I mostly have to apply package name changes in order to make our project compile. But I cannot get it to run.

1. At first I found out that `cucumber.properties` must no longer reside in the classpath root but now needs to be in the classpath under `io/cucumber/core/options/cucumber.properties`. Is that the expected behaviour? Otherwise `io.cucumber.core.options.CucumberProperties` line 25 should have a leading '/'.
2. My custom ObjectFactory cannot be loaded due to a CucumberException which states that it cannot be cast. The error is in `io.cucumber.core.options.CucumberPropertiesParser` line 53. It should be: `if (!ObjectFactory.class.isAssignableFrom(objectFactoryClass)) {`

I can fix both along with this issue or create separate issues (much more work). Which way would you have it?

**Comment by mpkorstanje** [comment]
Please do fix it straight away.

**Comment by ralphkar** [comment]
I'll provide a PR beginning of august. It's almost ready but I'm gone for a while. Please be patient.

**Comment by mpkorstanje** [comment]
No rush. I might fix both bugs myself and then v5 is good to go. Adding it to  `@CucumberOptions` would be a nice extra feature.

**Comment by mpkorstanje** [comment]
Marking this as 5.0.0 for now to fix the bugs.

**Comment by mpkorstanje** [comment]
Fixed.

**Comment by ralphkar** [comment]
Cool!
I basically have the CLI and annotation (jUnit and TestNG) stuff ready, but not tested. I will do extensive tests when I am back (in about 10 days). I suspect that the ServiceLoader stuff will make things more difficult. We'll see.

**Comment by mpkorstanje** [comment]
No worries. No rush. I've found the ServicceLoader to be easier to test/use then the classpath scanning. But let me know if there is anything you are stuck on when you get back.

**Comment by ralphkar** [comment]
Fixed in #1710
